src/grammars/peg.md

```diff
@@ -1,105 +1,148 @@
-concept to regular expressions with named groups with the exception that regex
-is *generative*, i.e. used to generate strings, while PEG is *analytic*, i.e.
-used to test whether strings match the grammar.
+Parsing expression grammars (PEGs) are simply a strict representation of the
+simple imperative code that you would write if you were writing a parser by
+hand.
 
-## Generative vs. analytics approach
+```
+digit = {           // To recognize a digit...
+    '0'..'9'        //   take any character from '0' to '9'.
+}
+expression = {      // To recognize an expression...
+    digit+          //   first take as many digits as possible (at least one)...
+    | "true"        //   or, if that fails, the string "true".
+}
+```
+
+In fact, `pest` produces code that is quite similar to the pseudo-code in the
+comments above.
 
-Let's take a look at a simple regex example in order to illustrate these
-concepts a bit better:
+## Eagerness
 
-```javascript
-/^(a*)(a)$/
+When a [repetition] PEG expression is run on an input string,
+
+```
+('0'..'9')+      // one or more characters from '0' to '9'
 ```
 
-This regex generates a string that starts from the beginning (`^`), matches zero
-or more `'a'` (`a*`), matches one `'a'` (`a`), then it makes sure it reached the
-end of its input. Since this a generative definition, we can generate the string
-`"aa"` by generating one `'a'` from the first group and the second `'a'` from
-the second group.
+it runs that expression as many times as it can (matching "eagerly", or
+"greedily"). It either succeeds, consuming whatever it matched and passing the
+remaining input on to the next step in the parser,
 
-If we define the same rule in pest, it will look something like this:
+```
+"42 boxes"
+ ^ Running ('0'..'9')+
 
+"42 boxes"
+   ^ Successfully took one or more digits!
+
+" boxes"
+ ^ Remaining unparsed input.
 ```
-soi ~ "a"* ~ "a" ~ eoi
+
+or fails, consuming nothing.
+
 ```
+"galumphing"
+ ^ Running ('0'..'9')+
+   Failed to take one or more digits!
 
-`soi` and `eoi` stand for start-of-input and end-of-input respectively and work
-much like their regex counterparts. `~` is used in order to crate sequences,
-while the rest of the syntax should be straightforward coming from the previous
-example.
+"galumphing"
+ ^ Remaining unparsed input (everything).
+```
 
-Part of the analytical approach of PEGs is the fact that they run eagerly. If we
-take the example generated by the regex, the PEG variant will eagerly match both
-`a`s in the `"a"*` expression, leaving the second `"a"` unsatisfied. Every
-expression in PEG tries to match as much as it possibly can. And, once it
-matched, it will not backtrack in order to satisfy the next expression. At
-first, this may seem like a bit of a hindrance, but it offers something else in
-exchange.
+If an expression fails to match, the failure propagates upwards, eventually
+leading to a failed parse. However, a failure can be "caught" using the *choice
```

Comment:
Although this is true, it's not the only way to "catch" the failures. Maybe a *slight* change in wording is due? ("One way to catch...")